sentry-graphql-java/src/main/java/io/sentry/graphql/SentryInstrumentation.java

```diff
@@ -0,0 +1,167 @@
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldFetchParameters;
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.GraphQLNonNull;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLOutputType;
+import io.sentry.HubAdapter;
+import io.sentry.IHub;
+import io.sentry.ISpan;
+import io.sentry.SpanStatus;
+import io.sentry.util.Objects;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class SentryInstrumentation extends SimpleInstrumentation {
+  private final @NotNull IHub hub;
+  private final @Nullable BeforeSpanCallback beforeSpan;
+
+  public SentryInstrumentation(
+      final @NotNull IHub hub, final @Nullable BeforeSpanCallback beforeSpan) {
+    this.hub = Objects.requireNonNull(hub, "hub is required");
+    this.beforeSpan = beforeSpan;
+  }
+
+  public SentryInstrumentation(final @NotNull IHub hub) {
+    this(hub, null);
+  }
+
+  public SentryInstrumentation() {
+    this(HubAdapter.getInstance());
+  }
+
+  @Override
+  public @NotNull InstrumentationState createState() {
+    return new TracingState();
+  }
+
+  @Override
+  public @NotNull InstrumentationContext<ExecutionResult> beginExecution(
+      final @NotNull InstrumentationExecutionParameters parameters) {
+    final TracingState tracingState = parameters.getInstrumentationState();
+    tracingState.setTransaction(hub.getSpan());
+    return super.beginExecution(parameters);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public @NotNull DataFetcher<?> instrumentDataFetcher(
+      final @NotNull DataFetcher<?> dataFetcher,
+      final @NotNull InstrumentationFieldFetchParameters parameters) {
+    // We only care about user code
+    if (parameters.isTrivialDataFetcher()) {
+      return dataFetcher;
+    }
+
+    return environment -> {
+      final TracingState tracingState = parameters.getInstrumentationState();
+      final ISpan transaction = tracingState.getTransaction();
+      if (transaction != null) {
+        final ISpan span = transaction.startChild(findDataFetcherTag(parameters));
+        try {
+          final Object result = dataFetcher.get(environment);
+          if (result instanceof CompletableFuture) {
+            ((CompletableFuture<?>) result)
+                .whenComplete(
+                    (r, ex) -> {
+                      if (ex != null) {
+                        span.setThrowable(ex);
+                        span.setStatus(SpanStatus.INTERNAL_ERROR);
+                      }
+                      finish(span, environment, r);
```

Comment:
Can we run into some concurrency issues since it's an async callback here?

Reply:
No, it's made thread safe.